# Sendy Logistics Challenge Regression Model¶
<br></br>



![Image of Yaktocat](https://cdn1.vc4a.com/media/2015/12/Sendy-delivery-900x322.jpg)


# Introduction

Sendy is a business-to-business platform established in 2014, to enable businesses of all types and sizes to transport goods more efficiently across East Africa.

Logistics is fundamental to the success of a business, and efficient and affordable logistics is a vital component of economic development. To make Sendy logistics affordable and effiecient, this notebooks predicts the estimated time of delivery of orders, from the point of driver pickup to the point of arrival at final destination. This prediction will enhance customer communication and improve the reliability of its service; which will ultimately improve customer experience. 

The solution provided by this notebook will enable Sendy to realise cost savings, and ultimately reduce the cost of doing business, through improved resource management and planning for order scheduling. The accurate arrival time prediction will help all businesses involved to improve their logistics and communicate an accurate time to their customers.

# Importing the Libraries

In [1]:
# 
import numpy as np 

# data processing
import pandas as pd 

# data visualization


# Getting the Data

In [2]:
# the csv files are converted to panda dataframe and are renamed.
Train_df= pd.read_csv("Train.csv")
Test_df= pd.read_csv("Test.csv")
Riders_df= pd.read_csv("Riders.csv")
VariableDefinitions_df= pd.read_csv("VariableDefinitions.csv")
SampleSubmission_df= pd.read_csv("SampleSubmission.csv")


# Data Exploration/Analysis

Exploring data sets helps in developing deep understanding about the data, the data set is Explored with imported libraries.

In [3]:
#Train data set
Train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21201 entries, 0 to 21200
Data columns (total 29 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   Order No                                   21201 non-null  object 
 1   User Id                                    21201 non-null  object 
 2   Vehicle Type                               21201 non-null  object 
 3   Platform Type                              21201 non-null  int64  
 4   Personal or Business                       21201 non-null  object 
 5   Placement - Day of Month                   21201 non-null  int64  
 6   Placement - Weekday (Mo = 1)               21201 non-null  int64  
 7   Placement - Time                           21201 non-null  object 
 8   Confirmation - Day of Month                21201 non-null  int64  
 9   Confirmation - Weekday (Mo = 1)            21201 non-null  int64  
 10  Confirmation - Time   

The training-set has 21201 examples and 28 features + the target variable (Time from Pickup to Arrival). 6 of the features are floats, 13 are integers and 10 are objects. Below is the list of the features with a short description:

In [4]:
Train_df.describe()

Platform Type  Placement - Day of Month  Placement - Weekday (Mo = 1)  \
count   21201.000000              21201.000000                  21201.000000   
mean        2.752182                 15.653696                      3.240083   
std         0.625178                  8.798916                      1.567295   
min         1.000000                  1.000000                      1.000000   
25%         3.000000                  8.000000                      2.000000   
50%         3.000000                 15.000000                      3.000000   
75%         3.000000                 23.000000                      5.000000   
max         4.000000                 31.000000                      7.000000   

       Confirmation - Day of Month  Confirmation - Weekday (Mo = 1)  \
count                 21201.000000                     21201.000000   
mean                     15.653837                         3.240225   
std                       8.798886                         1.567228   
min                       1.000000                         1.000000   
25%                       8.000000                         2.000000   
50%                      15.000000                         3.000000   
75%                      23.000000                         5.000000   
max                      31.000000                         7.000000   

       Arrival at Pickup - Day of Month  Arrival at Pickup - Weekday (Mo = 1)  \
count                      21201.000000                          21201.000000   
mean                          15.653837                              3.240225   
std                            8.798886                              1.567228   
min                            1.000000                              1.000000   
25%                            8.000000                              2.000000   
50%                           15.000000                              3.000000   
75%                           23.000000                              5.000000   
max                           31.000000                              7.000000   

       Pickup - Day of Month  Pickup - Weekday (Mo = 1)  \
count           21201.000000               21201.000000   
mean               15.653837                   3.240225   
std                 8.798886                   1.567228   
min                 1.000000                   1.000000   
25%                 8.000000                   2.000000   
50%                15.000000                   3.000000   
75%                23.000000                   5.000000   
max                31.000000                   7.000000   

       Arrival at Destination - Day of Month  \
count                           21201.000000   
mean                               15.653837   
std                                 8.798886   
min                                 1.000000   
25%                                 8.000000   
50%                                15.000000   
75%                                23.000000   
max                                31.000000   

       Arrival at Destination - Weekday (Mo = 1)  Distance (KM)   Temperature  \
count                               21201.000000   21201.000000  16835.000000   
mean                                    3.240225       9.506533     23.258889   
std                                     1.567228       5.668963      3.615768   
min                                     1.000000       1.000000     11.200000   
25%                                     2.000000       5.000000     20.600000   
50%                                     3.000000       8.000000     23.500000   
75%                                     5.000000      13.000000     26.000000   
max                                     7.000000      49.000000     32.100000   

       Precipitation in millimeters    Pickup Lat   Pickup Long  \
count                    552.000000  21201.000000  21201.000000   
mean                       7.905797     -1.281470     36.811264   
std                 

from the  above, it is observable that distance from pickup to drop off location is from 1km to 49km and that on average the Time from Pickup to Arrival is ranging from 1sec 7883 seconds.

In [5]:
Train_df.head(8)

Order No       User Id Vehicle Type  Platform Type  \
0   Order_No_4211   User_Id_633         Bike              3   
1  Order_No_25375  User_Id_2285         Bike              3   
2   Order_No_1899   User_Id_265         Bike              3   
3   Order_No_9336  User_Id_1402         Bike              3   
4  Order_No_27883  User_Id_1737         Bike              1   
5   Order_No_7408  User_Id_1342         Bike              3   
6  Order_No_22680  User_Id_2803         Bike              3   
7  Order_No_21578  User_Id_1075         Bike              3   

  Personal or Business  Placement - Day of Month  \
0             Business                         9   
1             Personal                        12   
2             Business                        30   
3             Business                        15   
4             Personal                        13   
5             Business                        14   
6             Business                         9   
7             Business                        11   

   Placement - Weekday (Mo = 1) Placement - Time  Confirmation - Day of Month  \
0                             5       9:35:46 AM                            9   
1                             5      11:16:16 AM                           12   
2                             2      12:39:25 PM                           30   
3                             5       9:25:34 AM                           15   
4                             1       9:55:18 AM                           13   
5                             5       3:07:35 PM                           14   
6                             5       9:33:45 AM                            9   
7                             1       2:13:01 PM                           11   

   Confirmation - Weekday (Mo = 1)  ... Arrival at Destination - Time  \
0                                5  ...                   10:39:55 AM   
1                                5  ...                   12:17:22 PM   
2                                2  ...                    1:00:38 PM   
3                                5  ...                   10:05:27 AM   
4                                1  ...                   10:25:37 AM   
5                                5  ...                    4:23:41 PM   
6                                5  ...                   10:19:45 AM   
7                                1  ...                    2:33:26 PM   

   Distance (KM)  Temperature Precipitation in millimeters  Pickup Lat  \
0              4         20.4                          NaN   -1.317755   
1             16         26.4                          NaN   -1.351453   
2              3          NaN                          NaN   -1.308284   
3              9         19.2                          NaN   -1.281301   
4              9         15.4                          NaN   -1.266597   
5              9         27.2                          NaN   -1.302583   
6              5         20.3                          NaN   -1.279395   
7              3         28.7                          NaN   -1.252796   

   Pickup Long Destination Lat  Destination Long      Rider Id  \
0    36.830370       -1.300406         36.829741  Rider_Id_432   
1    36.899315       -1.295004         36.814358  Rider_Id_856   
2    36.843419       -1.300921         36.828195  Rider_Id_155   
3    36.832396       -1.257147         36.795063  Rider_Id_855   
4    36.792118       -1.295041         36.809817  Rider_Id_770   
5    36.767081       -1.257309         36.806008  Rider_Id_124   
6    36.825364       -1.276574         36.851365  Rider_Id_114   
7    36.800313       -1.255189         36.782203  Rider_Id_913   

  Time from Pickup to Arrival  
0                         745  
1                        1993  
2                         455  
3                        1341  
4                        1214  
5                        3191  
6                        1380  
7                         646  

[8 rows x 29 columns]

From the table above,few things can be noted. some of the features need to be converted into numeric ones, so that the machine learning algorithms can process them. Furthermore. Some more features, that contain missing values (NaN = not a number) can also be spoted, that should be deal with.

__more detailed look at what data is actually missing:__

In [6]:
tr_total = Train_df.isnull().sum().sort_values(ascending=False)
tr_percent_1 = Train_df.isnull().sum()/Train_df.isnull().count()*100
tr_percent_2 = (round(tr_percent_1, 1)).sort_values(ascending=False)
tr_missing_data = pd.concat([tr_total, tr_percent_2], axis=1, keys=['Total', '%'])
tr_missing_data.head(5)

Total     %
Precipitation in millimeters          20649  97.4
Temperature                            4366  20.6
Time from Pickup to Arrival               0   0.0
Arrival at Pickup - Weekday (Mo = 1)      0   0.0
User Id                                   0   0.0

It will be a bit tricky to deal with the 'Temparature' feature, which has 4366 missing values. The 'Precipitation in mm' feature needs further investigation, but it looks like it would be better if it is dropped off from the dataset, since 97.4 % of it are missing.

In [7]:
Train_df.columns.values

array(['Order No', 'User Id', 'Vehicle Type', 'Platform Type',
       'Personal or Business', 'Placement - Day of Month',
       'Placement - Weekday (Mo = 1)', 'Placement - Time',
       'Confirmation - Day of Month', 'Confirmation - Weekday (Mo = 1)',
       'Confirmation - Time', 'Arrival at Pickup - Day of Month',
       'Arrival at Pickup - Weekday (Mo = 1)', 'Arrival at Pickup - Time',
       'Pickup - Day of Month', 'Pickup - Weekday (Mo = 1)',
       'Pickup - Time', 'Arrival at Destination - Day of Month',
       'Arrival at Destination - Weekday (Mo = 1)',
       'Arrival at Destination - Time', 'Distance (KM)', 'Temperature',
       'Precipitation in millimeters', 'Pickup Lat', 'Pickup Long',
       'Destination Lat', 'Destination Long', 'Rider Id',
       'Time from Pickup to Arrival'], dtype=object)

Above are the 28 features + the target variable (Time from Pickup to Arrival). Platform Type, Personal or Business and Distance (KM) would be correlated with Time from Pickup to Arrival.

In [8]:
# Test data set
#the data set is similar to the train set.
Test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7068 entries, 0 to 7067
Data columns (total 25 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   Order No                              7068 non-null   object 
 1   User Id                               7068 non-null   object 
 2   Vehicle Type                          7068 non-null   object 
 3   Platform Type                         7068 non-null   int64  
 4   Personal or Business                  7068 non-null   object 
 5   Placement - Day of Month              7068 non-null   int64  
 6   Placement - Weekday (Mo = 1)          7068 non-null   int64  
 7   Placement - Time                      7068 non-null   object 
 8   Confirmation - Day of Month           7068 non-null   int64  
 9   Confirmation - Weekday (Mo = 1)       7068 non-null   int64  
 10  Confirmation - Time                   7068 non-null   object 
 11  Arrival at Pickup

The training-set has 7068 entries and 23 features + the target variable (Time from Pickup to Arrival). 6 of the features are floats, 10 are integers and 9 are objects. 
Below is the list of the features with a short description:

In [9]:
Train_df.describe()

Platform Type  Placement - Day of Month  Placement - Weekday (Mo = 1)  \
count   21201.000000              21201.000000                  21201.000000   
mean        2.752182                 15.653696                      3.240083   
std         0.625178                  8.798916                      1.567295   
min         1.000000                  1.000000                      1.000000   
25%         3.000000                  8.000000                      2.000000   
50%         3.000000                 15.000000                      3.000000   
75%         3.000000                 23.000000                      5.000000   
max         4.000000                 31.000000                      7.000000   

       Confirmation - Day of Month  Confirmation - Weekday (Mo = 1)  \
count                 21201.000000                     21201.000000   
mean                     15.653837                         3.240225   
std                       8.798886                         1.567228   
min                       1.000000                         1.000000   
25%                       8.000000                         2.000000   
50%                      15.000000                         3.000000   
75%                      23.000000                         5.000000   
max                      31.000000                         7.000000   

       Arrival at Pickup - Day of Month  Arrival at Pickup - Weekday (Mo = 1)  \
count                      21201.000000                          21201.000000   
mean                          15.653837                              3.240225   
std                            8.798886                              1.567228   
min                            1.000000                              1.000000   
25%                            8.000000                              2.000000   
50%                           15.000000                              3.000000   
75%                           23.000000                              5.000000   
max                           31.000000                              7.000000   

       Pickup - Day of Month  Pickup - Weekday (Mo = 1)  \
count           21201.000000               21201.000000   
mean               15.653837                   3.240225   
std                 8.798886                   1.567228   
min                 1.000000                   1.000000   
25%                 8.000000                   2.000000   
50%                15.000000                   3.000000   
75%                23.000000                   5.000000   
max                31.000000                   7.000000   

       Arrival at Destination - Day of Month  \
count                           21201.000000   
mean                               15.653837   
std                                 8.798886   
min                                 1.000000   
25%                                 8.000000   
50%                                15.000000   
75%                                23.000000   
max                                31.000000   

       Arrival at Destination - Weekday (Mo = 1)  Distance (KM)   Temperature  \
count                               21201.000000   21201.000000  16835.000000   
mean                                    3.240225       9.506533     23.258889   
std                                     1.567228       5.668963      3.615768   
min                                     1.000000       1.000000     11.200000   
25%                                     2.000000       5.000000     20.600000   
50%                                     3.000000       8.000000     23.500000   
75%                                     5.000000      13.000000     26.000000   
max                                     7.000000      49.000000     32.100000   

       Precipitation in millimeters    Pickup Lat   Pickup Long  \
count                    552.000000  21201.000000  21201.000000   
mean                       7.905797     -1.281470     36.811264   
std                 

It is observable that distance from pickup to drop off location is also from 1.0km to 49.0km and that on average the Time from Pickup to Arrival is ranging from 1sec 7883 seconds.

__more detailed look at what data is actually missing:__

In [10]:
ts_total = Test_df.isnull().sum().sort_values(ascending=False)
ts_percent_1 = Test_df.isnull().sum()/Test_df.isnull().count()*100
ts_percent_2 = (round(ts_percent_1, 1)).sort_values(ascending=False)
ts_missing_data = pd.concat([ts_total, ts_percent_2], axis=1, keys=['Total', '%'])
ts_missing_data.head(5)

Total     %
Precipitation in millimeters   6869  97.2
Temperature                    1437  20.3
Rider Id                          0   0.0
Confirmation - Time               0   0.0
User Id                           0   0.0

the missing data is pretty much similar the thr train set .

In [11]:
# Rider data set
Riders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 960 entries, 0 to 959
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Rider Id        960 non-null    object 
 1   No_Of_Orders    960 non-null    int64  
 2   Age             960 non-null    int64  
 3   Average_Rating  960 non-null    float64
 4   No_of_Ratings   960 non-null    int64  
dtypes: float64(1), int64(3), object(1)
memory usage: 37.6+ KB


There are 960 entries and 5 features. features consist of  a float, 3 int and 1 object.

In [12]:
Riders_df.describe()

No_Of_Orders          Age  Average_Rating  No_of_Ratings
count    960.000000   960.000000      960.000000     960.000000
mean    1864.851042  1200.234375       13.412604     427.983333
std     1880.337785   810.930171        2.675794     486.957931
min        2.000000    96.000000        0.000000       0.000000
25%      261.750000   478.250000       13.500000      30.000000
50%     1475.500000  1021.000000       14.000000     223.000000
75%     2847.250000  1891.500000       14.300000     678.750000
max     9756.000000  3764.000000       15.200000    2298.000000

__more detailed look at what data is actually missing:__

In [13]:
r_total = Riders_df.isnull().sum().sort_values(ascending=False)
r_percent_1 = Riders_df.isnull().sum()/Riders_df.isnull().count()*100
r_percent_2 = (round(r_percent_1, 1)).sort_values(ascending=False)
r_missing_data = pd.concat([r_total, r_percent_2], axis=1, keys=['Total', '%'])
r_missing_data.head(5)

Total    %
No_of_Ratings       0  0.0
Average_Rating      0  0.0
Age                 0  0.0
No_Of_Orders        0  0.0
Rider Id            0  0.0

there are no missing values in the riders data set.

In [14]:
# VariableDefinition data set
VariableDefinitions_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35 entries, 0 to 34
Data columns (total 2 columns):
 #   Column                               Non-Null Count  Dtype 
---  ------                               --------------  ----- 
 0   Order No                             34 non-null     object
 1   Unique number identifying the order  33 non-null     object
dtypes: object(2)
memory usage: 688.0+ bytes


This data set has the descriptive details about all the features which are in the Train,Test and Rides data set.

In [15]:
          VariableDefinitions_df.head(8)

Order No  \
0                       User Id   
1                  Vehicle Type   
2                 Platform Type   
3          Personal or Business   
4      Placement - Day of Month   
5  Placement - Weekday (Mo = 1)   
6              Placement - Time   
7   Confirmation - Day of Month   

                 Unique number identifying the order  
0  Unique number identifying the customer on a pl...  
1  For this competition limited to bikes, however...  
2  Platform used to place the order, there are 4 ...  
3                                      Customer type  
4                  Placement - Day of Month i.e 1-31  
5                   Placement - Weekday (Monday = 1)  
6  Placement - Time - Time of day the order was p...  
7               Confirmation - Day of Month i.e 1-31

In [16]:
SampleSubmission_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7068 entries, 0 to 7067
Data columns (total 2 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Order_No                     7068 non-null   object 
 1   Time from Pickup to Arrival  3 non-null      float64
dtypes: float64(1), object(1)
memory usage: 110.6+ KB


In [17]:
SampleSubmission_df.head(8)

Order_No  Time from Pickup to Arrival
0  Order_No_19248                        567.0
1  Order_No_12736                       4903.0
2    Order_No_768                       5649.0
3  Order_No_15332                          NaN
4  Order_No_21373                          NaN
5  Order_No_14573                          NaN
6   Order_No_6731                          NaN
7  Order_No_18436                          NaN

# Data Preprocessing

Riders set will be joined with the train set and everything execpt 'Distance(KM),Age, Average ratings and Time from Pickup to Arrival' will be dropped from the train. This is because they do not contribute to Time from Pickup to Arrival. In the Test set, Order_No will not droped from the test set, since it is required there for the submission. Riders will be joined to the test set awell and same features will be dropped.

In [18]:
Train_df = pd.merge(Train_df, Riders_df, left_on='Rider Id', right_on='Rider Id', how='left').drop(["Order No","User Id", 
        "Rider Id",'Vehicle Type', 'Platform Type', "Personal or Business", 
        "Arrival at Pickup - Day of Month", "Arrival at Pickup - Weekday (Mo = 1)", "Arrival at Pickup - Time",
        "Pickup - Day of Month", "Pickup - Weekday (Mo = 1)",
        "Pickup - Time", "Arrival at Destination - Day of Month", 
        "Arrival at Destination - Weekday (Mo = 1)", 
        "Arrival at Destination - Time","Placement - Day of Month",
        "Placement - Weekday (Mo = 1)", "Placement - Time", 
        "Confirmation - Day of Month", "Confirmation - Weekday (Mo = 1)", 
        "Confirmation - Time", "Temperature", "Precipitation in millimeters", 
        "Pickup Lat", "Pickup Long", "Destination Lat", "Destination Long", 
        "No_Of_Orders","No_of_Ratings"], axis=1)

In [19]:
# A look at training set before start training the models.
Train_df.head(8)

Distance (KM)  Time from Pickup to Arrival   Age  Average_Rating
0              4                          745  1309            13.8
1             16                         1993   339            13.6
2              3                          455   242            12.5
3              9                         1341   283            14.5
4              9                         1214   872            14.1
5              9                         3191  1204            14.2
6              5                         1380   556            13.7
7              3                          646  2119            14.4

In [20]:
Test_df = pd.merge(Test_df, Riders_df, left_on='Rider Id', right_on='Rider Id', how='left').drop(['Vehicle Type', 
        'Platform Type', 'Personal or Business', 'Placement - Day of Month',
       'Placement - Weekday (Mo = 1)', 'Placement - Time','Confirmation - Day of Month', 
       'Confirmation - Weekday (Mo = 1)','Confirmation - Time', 'Arrival at Pickup - Day of Month',
       'Arrival at Pickup - Weekday (Mo = 1)', 'User Id', 'No_Of_Orders',
       'Arrival at Pickup - Time','Pickup - Day of Month', 'Pickup - Weekday (Mo = 1)',
       'Pickup - Time', 'Temperature','Precipitation in millimeters', 'Pickup Lat', 'Pickup Long',
       'Destination Lat', 'Destination Long', 'Rider Id',"No_of_Ratings"], axis = 1) 

In [21]:
Test_df.head(10)

Order No  Distance (KM)   Age  Average_Rating
0  Order_No_19248              8  1511            13.3
1  Order_No_12736              5   273            14.4
2    Order_No_768              5   460            13.1
3  Order_No_15332              5   560            13.7
4  Order_No_21373              6  1120            14.1
5  Order_No_14573             16   741            14.0
6   Order_No_6731             18  1113            14.4
7  Order_No_18436              8  1151            13.5
8   Order_No_2288              8   286            13.6
9   Order_No_9063             15   635            14.3

# Building Machine Learning Models